aggregator/src/aggregator/queue.rs

```diff
@@ -0,0 +1,570 @@
+                    max_concurrency: concurrency,
+                },
+            ));
+
+            join_all((0..requests).map(|_| {
+                let handler = Arc::clone(&handler);
+                async move {
+                    get("/").run_async(&handler).await;
+                }
+            }))
+            .await;
+        });
+    }
+
+    #[quickcheck]
+    fn quickcheck_lifo_cancel(parameters: Parameters) {
+        install_test_trace_subscriber();
+        debug!(?parameters, "quickcheck_lifo_cancel parameters");
+        let Parameters {
+            runtime_flavor,
+            depth,
+            concurrency,
+        } = parameters;
+
+        runtime_flavor.build().block_on(async move {
+            let unhang = Arc::new(Semaphore::new(0));
+            let queue = Arc::new(LIFORequestQueue::new(concurrency, depth).unwrap());
+            let handler = Arc::new(queued_lifo(
+                Arc::clone(&queue),
+                HangingHandler {
+                    unhang: Arc::clone(&unhang),
+                },
+            ));
+
+            let requests =
+                fill_queue(Arc::clone(&queue), Arc::clone(&handler), concurrency, depth).await;
+
+            // Abort all outstanding requests.
+            requests.iter().for_each(|request| request.abort());
+
+            // Wait until the queue is empty.
+            while queue.len() > 0 {
+                yield_now().await;
+            }
+
+            unhang.close();
+            let request = get("/").run_async(&handler).await;
+            assert_ok!(request);
+
+            for handle in requests {
+                // These handles will return a JoinError, but we're moreso interested to see if
+                // they've all terminated.
+                let _ = handle.await;
+            }
+        });
+    }
+
+    #[quickcheck]
+    fn quickcheck_lifo_full(parameters: Parameters) {
+        install_test_trace_subscriber();
+        debug!(?parameters, "quickcheck_lifo_full parameters");
+        let Parameters {
+            runtime_flavor,
+            depth,
+            concurrency,
+        } = parameters;
+
+        runtime_flavor.build().block_on(async move {
+            let unhang = Arc::new(Semaphore::new(0));
+            let queue = Arc::new(LIFORequestQueue::new(concurrency, depth).unwrap());
+            let handler = Arc::new(queued_lifo(
+                Arc::clone(&queue),
+                HangingHandler {
+                    unhang: Arc::clone(&unhang),
+                },
+            ));
+
+            let requests =
+                fill_queue(Arc::clone(&queue), Arc::clone(&handler), concurrency, depth).await;
+
+            let request = get("/").run_async(&handler).await;
+            assert_status!(request, Status::ServiceUnavailable);
+
+            unhang.close();
+
+            // Let the dispatcher get a turn on the scheduler.
+            while queue.len() == depth {
+                yield_now().await;
```

Comment:
Similarly to above, using yield_now concedes that these tests are somewhat non-deterministic wrt to the choices of the tokio scheduler. However, I think this plus quickcheck yields more representative, if flaky, tests.

The alternative is pulling out the dispatcher sweep logic into a function we can invoke in tests, but that amount of determinism is not available in reality. (Indeed, by writing tests like this I've uncovered at least one bug.) But I'm open to input on this.

Reply:
Most yield_now() points were eliminated in my latest iteration, mostly by virtue of needing an await point to read the queue depth.